File: CPP/Lesson_142_binary_tree_right_side_view.cpp
// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 142 -- Binary Tree Right Side View
// Category   : Trees
// Difficulty : Medium
// Study Plan : Day 71
// =============================================================
//
// QUESTION:
//   Return values seen from the right side of a binary tree.
// =============================================================
#include <vector>
#include <string>
#include <iostream>
#include <stack>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <map>
#include <set>
#include <algorithm>
#include <climits>
#include <numeric>
#include <functional>
#include <cmath>
using namespace std;
struct TN{int val;TN*l;TN*r;TN(int v,TN*a=nullptr,TN*b=nullptr):val(v),l(a),r(b){}};
vector<int> view(TN* root){vector<int> res;if(!root)return res;queue<TN*> q;q.push(root);while(!q.empty()){int n=q.size();for(int i=0;i<n;i++){auto x=q.front();q.pop();if(i==n-1)res.push_back(x->val);if(x->l)q.push(x->l);if(x->r)q.push(x->r);}}return res;}
int main(){TN n5(5),n4(4),n2(2,nullptr,&n5),n3(3,nullptr,&n4),n1(1,&n2,&n3);auto r=view(&n1);for(int x:r)cout<<x<<" ";cout<<"\n";}

File: Python/Lesson_142_binary_tree_right_side_view.py
# =============================================================
# MIT License | @analyticswithharry2026
# GitHub  : https://github.com/analyticswithharry
# YouTube : Analytics with Harry
# =============================================================
# Lesson     : 142 -- Binary Tree Right Side View
# Category   : Trees
# Difficulty : Medium
# Study Plan : Day 71
# =============================================================
#
# QUESTION:
#   Return values seen from the right side of a binary tree.
# =============================================================
from collections import deque
class TN:
    def __init__(self,v,l=None,r=None): self.val=v; self.left=l; self.right=r
def view(root):
    if not root: return []
    q=deque([root]); res=[]
    while q:
        n=len(q)
        for i in range(n):
            x=q.popleft()
            if i==n-1: res.append(x.val)
            if x.left: q.append(x.left)
            if x.right: q.append(x.right)
    return res

if __name__=="__main__":
    r=TN(1,TN(2,None,TN(5)),TN(3,None,TN(4)))
    print(view(r))
